`unneeded_escaping` false positive in 0.63.0

### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description


(fairly) minimal example I created:

```swift
import Foundation

public class Uploader {

    public func start(completion: @escaping (Result<String, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            let result: [Result<String, Error>] = [1].map { _ in .success("Success") }
            completion(result.first!)
        }
    }
}

```

Running:

```bash
swiftlint lint --only-rule unneeded_escaping --no-cache ./a.swift
```

Results in:

```
a.swift:5:35: warning: Unneeded Escaping Violation: @escaping attribute not required as 'completion' does not escape (unneeded_escaping)
```

but it cannot be removed:

```bash
swift ./a.swift                                                  
a.swift:6:57: error: escaping closure captures non-escaping parameter 'completion'
```

It somehow has todo with the map, because if I change the code to `let result = Result.success("Success")`, no error is reported anymore

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,7 +16,10 @@
 
 ### Bug Fixes
 
-* None.
+* Fix false positive in `unneeded_escaping` rule when an escaping closure is used in
+  a nested closure preceded by another closure.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#6410](https://github.com/realm/SwiftLint/issues/6410)
 
 ## 0.63.0: High-Speed Extraction
 

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededEscapingRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededEscapingRule.swift
@@ -87,7 +87,7 @@ private extension UnneededEscapingRule {
 private final class EscapeChecker: SyntaxVisitor {
     var taintedVariables = Set<String>()
     var doesEscape = false
-    var inClosureContext = false
+    var inClosureContext = Stack<Bool>()
     let isAutoclosure: Bool
 
     init(paramName: String, isAutoclosure: Bool) {
@@ -138,19 +138,19 @@ private final class EscapeChecker: SyntaxVisitor {
     }
 
     override func visit(_: ClosureExprSyntax) -> SyntaxVisitorContinueKind {
-        inClosureContext = true
+        inClosureContext.push(true)
         return .visitChildren
     }
 
     override func visitPost(_: ClosureExprSyntax) {
-        inClosureContext = false
+        inClosureContext.pop()
     }
 
     override func visitPost(_ node: DeclReferenceExprSyntax) {
         guard isTainted(ExprSyntax(node)) else {
             return
         }
-        if inClosureContext || [.arrayElement, .dictionaryElement].contains(node.parent?.kind) {
+        if (inClosureContext.peek() ?? false) || [.arrayElement, .dictionaryElement].contains(node.parent?.kind) {
             doesEscape = true
         }
     }

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededEscapingRuleExamples.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededEscapingRuleExamples.swift
@@ -6,6 +6,14 @@ struct UnneededEscapingRuleExamples {
         func outer(completion: @escaping () -> Void) { inner(completion: completion) }
         """),
         Example("""
+        func f(completion: @escaping [Int] -> Void) {
+            g {
+                let result = [1].map { _ in 0 }
+                completion(result)
+            }
+        }
+        """),
+        Example("""
         func outer(closure: @escaping @autoclosure () -> String) {
             inner(closure: closure())
         }

